internal/tofu/transform_provider.go

```diff
@@ -102,147 +103,96 @@ func (t *ProviderTransformer) Transform(g *Graph) error {
 
 	var diags tfdiags.Diagnostics
 
-	// To start, we'll collect the _requested_ provider addresses for each
-	// node, which we'll then resolve (handling provider inheritance, etc) in
-	// the next step.
-	// Our "requested" map is from graph vertices to string representations of
-	// provider config addresses (for deduping) to requests.
-	type ProviderRequest struct {
-		Addr  addrs.AbsProviderConfig
-		Exact bool // If true, inheritance from parent modules is not attempted
-	}
-	requested := map[dag.Vertex]map[string]ProviderRequest{}
-	needConfigured := map[string]addrs.AbsProviderConfig{}
+	// Now we'll go through all the requested addresses we just collected and
+	// figure out which _actual_ config address each belongs to, after resolving
+	// for provider inheritance and passing.
+	m := providerVertexMap(g)
 	for _, v := range g.Vertices() {
-		// Does the vertex _directly_ use a provider?
 		if pv, ok := v.(GraphNodeProviderConsumer); ok {
```

Comment:
let's reduce nesting here:

```suggestion
		if pv, ok := v.(GraphNodeProviderConsumer); !ok {
			continue
		}
		...
```

Reply:
We use pv below and can't omit the value here.

Reply:
[1a5e1ec](https://github.com/opentofu/opentofu/pull/2069/commits/1a5e1ec2d8b9938424e22ce2ff24936c96a769d4)